Add 'pre_page_moved' and 'post_page_moved' signals

ababic opened the pull request:
Supersedes #4975

A client project I've been working on recently could really benefit from these changes, so I've gone ahead and implemented myself. Apologies to you, @joncombe, if you'd already been looking at this.

* Do the tests still pass? Yes
* Does the code comply with the style guide? Yes
* For Python changes: Have you added tests to cover the new/fixed behaviour? Yes
* For front-end changes: Did you test on all of Wagtail’s supported browsers? N/A
* For new features: Has the documentation been updated accordingly? Yes

ababic commented:
It's a no-brainer to me that `wagtail.api` should utilise this signal to help purge pages from the API cache (in addition to the `page_published` and `page_unpublished` signals). However, doing so resulted in a few instances of the following error:

```
File "/home/travis/build/wagtail/wagtail/wagtail/api/v2/signal_handlers.py", line 15, in purge_page_from_cache
purge_url_from_cache(base_url + reverse('wagtailapi_v2:pages:detail', args=(instance.id, )))
TypeError: unsupported operand type(s) for +: 'NoneType' and 'str'
```

I initially tried to mitigate this by [raising a warning here](https://github.com/wagtail/wagtail/blob/master/wagtail/api/v2/signal_handlers.py#L14) when `WAGTAILAPI_BASE_URL` isn't set, instead of allowing this exception to be raised. But, I'm having doubts about whether that is really a good approach to be taking, as warnings output in test feedback is only marginally better. There also seems to be some co-dependency between `wagtail.api` and `wagtail.contrib.frontend_cache`, which I'm not quite clear on, so it's difficult to know what the consequences of this would be.

@kaedroho Do you have any thoughts on this?

tomdyson commented on wagtail/admin/views/pages.py:
```diff
@@ -792,7 +792,7 @@ def move_confirm(request, page_to_move_id, destination_id):
 
 def set_page_position(request, page_to_move_id):
-    page_to_move = get_object_or_404(Page, id=page_to_move_id)
+    page_to_move = get_object_or_404(Page, id=page_to_move_id).specific
```
Why do we need `specific` here?

ababic replied:
Part of the problem the client is having is that they are subscribing to the `post_save` signal using their specific page types as the `sender`, with the expectation that moving will result in receipt of the signal.

While `Page.move()` does in-turn call `Page.save()`, resulting in a `post_save` signal being emitted, because we are only dealing with a vanilla `Page` object here, that signal is only emitted with `Page` as the sender, and not the fully-fledged page type.

ababic replied:
The current behaviour is inconsistent with the `move_page()` view just above this, which does ultimately result in a `post_save` signal being emitted for the specific page type.

ababic replied:
@tomdyson / @kaedroho Have removed this change for now to simplify reviewing

tomdyson commented on docs/reference/signals.rst:
```diff
@@ -82,3 +82,17 @@ This signal is emitted from a ``Page`` when the page is unpublished.
+:sender: The page ``class``
+:instance: The specific ``Page`` instance.
+:position_page: The ``Page`` instance used in conjunction with ``position`` to specify the new tree position for ``instance`` (the intended parent or sibling page).
+:position: A string used in conjunction with ``position_page`` to specify where the new tree position for ``instance``. Common values are `'last-child'` if moving to a different branch, or `'left'` or `'right'` when reordering.
```
"to specify where the new tree position" -> "to specify the new tree position"

kaedroho commented on wagtail/contrib/frontend_cache/signal_handlers.py:
```diff
@@ -1,7 +1,11 @@
+from wagtail.core.signals import page_moved, page_published, page_unpublished
+
+
+def page_moved_signal_handler(instance, **kwargs):
```
Worth mentioning this separately in release notes.

ababic replied:
> I think we should also purge the URL of the page in it's previous location too.

That would be cool. How do you envisage this being supplied to signal handlers? would supplying `old_url_path` (defined currently in `Page.move()`) be enough here?

ababic replied:
@kaedroho I just realised something quite important. If a page has descendants, the paths/URLs of those descendants might also change as a result of `move()`. So, if we want to clear all outdated URLs from the cache, we really need the old URLs for all of those descendants, as well as that of the current page. Is it enough to provide a list of all outdated page URLs, or do you need the full page objects (in their 'before' state) in order to reliably clear things from the front-end cache?

ababic commented on docs/reference/signals.rst:
```diff
@@ -82,3 +82,17 @@ This signal is emitted from a ``Page`` when the page is unpublished.
+
+:sender: The page ``class``
+:instance: The specific ``Page`` instance.
+:position_page: The ``Page`` instance used in conjunction with ``position`` to specify the new tree position for ``instance`` (the intended parent or sibling page).
```
Thanks! I actually went for the terminology used in the view here, rather than by the `move()` method itself (so, there is already some consistency). If you can confirm that you'd be happier with `target`, I will happily change :)

kaedroho replied:
We probably should be matching the terminology with the ``move`` method since that's already public API. I'm happier with ``target``, thanks!

ababic commented on wagtail/admin/views/pages.py:
```diff
@@ -792,7 +792,7 @@ def move_confirm(request, page_to_move_id, destination_id):
 
 def set_page_position(request, page_to_move_id):
-    page_to_move = get_object_or_404(Page, id=page_to_move_id)
+    page_to_move = get_object_or_404(Page, id=page_to_move_id).specific
```
@kaedroho Please see above comments. This is more to do with resolving some inconsistency between this view and the `move_confirm()` view just above it.

ababic commented on docs/reference/signals.rst:
```diff
@@ -82,3 +82,17 @@ This signal is emitted from a ``Page`` when the page is unpublished.
 :kwargs: Any other arguments passed to ``page_unpublished.send()``
+
+
+page_moved
```
Good shout! I'll update the existing headings too, for consistency

ababic commented on docs/reference/signals.rst:
```diff
@@ -82,3 +82,17 @@ This signal is emitted from a ``Page`` when the page is unpublished.
+
+.. versionadded:: 2.5
+
+This signal is emitted from a ``Page`` when the page is moved to a different branch of the page tree, or reordered within the same branch.
```
Understood. I'll go with "section" :)

ababic commented on wagtail/admin/tests/test_pages_views.py:
```diff
@@ -2633,6 +2634,25 @@ def test_page_move_confirm(self):
+    def test_move_triggers_signal(self):
+        # Connect a mock signal handler to page_moved signal
+        mock_handler = mock.MagicMock()
+        page_moved.connect(mock_handler)
```
@kaedroho This is sensible, thanks. Should I update the other tests in this module while I'm here, or should that be done in a separate PR?

ababic commented:
@kaedroho A few more questions to you above :) If you also have any thoughts on the best way to resolve [the above](https://github.com/wagtail/wagtail/pull/5169#issuecomment-475962583) too, that would be great :)

ababic commented:
@kaedroho I've decided to scale back these changes to help with review. I'm going to remove the front-end cache integration for now to reduce sprawl, and just focus on making this signal as 'generically' helpful as possible (in a way that we might improve later)

kaedroho commented:
I agree, it's probably going to be tricky to do the cache stuff especially since we don't support wildcard cash invalidation yet so purging the subtree could be quite a lot of work in some cases.

We should come back to that bit once we've added support for wildcard invalidation (or, at least, async invalidation)

ababic commented:
@kaedroho I believe I've now addressed all of the points raised so far. 

I'm interested to hear what you think about the addition of `changed_url_paths` though. I was trying to strike a balance between supplying all relevant information, while also keeping the number of additional queries to a minimum. 

Adding the `ChangedPageURLPath` class here might possibly be considered overkill, but I figured it'd be convenient if we ever want to build on this (e.g. add helper methods to assist with urls / full url generation)

ababic commented:
Related issue: #2728

ababic commented:
I've since realised that `changed_url_paths` here would not actually be very useful in reliably determining old url/full url path paths, because to do that reliably, you need to use `reverse('wagtail_serve')`, and that will only work if the entire database state reflects the point in time you are interested in... if all the `path` and `url_path` values have already changed (which they will have by the time this signal is received), you won't be able to work out the old urls any longer.

I think the sensible solution here is to instead have two signals:

- `pre_page_move()` will be emitted before the database state changes, allowing you to derive **old** URLs if you need to.
- `post_page_move()` will be emitted after the database state changes, allowing you to derive **new** URLs if you need to.

ababic commented:
This is now ready for review :)

kaedroho commented:
Sorry for catching this a little late. Could these be renamed to `pre_page_move` and `post_page_move` (drop the 'd'). This is more consistent with Django's wording.

Also, do you think it sounds better to have the noun at the end? `pre_move_page`

ababic commented:
@kaedroho Yes, I'd agree that's better way to go. Should be sorted now :)

lb- commented:
@ababic  - thanks so much for this epic work, this has now been merged in via 3fe8775024c4ed5b78f69bd7d3f3fe05a1bafe2d
